### PR TITLE
CacheKey calculation and Injection

### PR DESCRIPTION
**Pull request by ryanai3**
added tasks to calculate cacheKeys and inject them.

### CODE REVIEW

**Comment by ryanai3** [comment]
@Jessie - please review

**Comment by ryanai3** [comment]
@jkinkead please review

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 1]
```diff
@@ -122,6 +124,11 @@ object DeployPlugin extends AutoPlugin {
     IO.delete((UniversalPlugin.autoImport.stagingDirectory in Universal).value)
   }
 
+  val stageAndCacheKeyTask = stageAndCacheKey := {
+    VersionInjectorPlugin.injectCacheKeyTask
+    println("") // so that we can be a Task[Unit]
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: This should be inferred without needing a special block.

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 2]
```diff
@@ -145,6 +152,7 @@ object DeployPlugin extends AutoPlugin {
   lazy val deployTask = deploy := {
     // Dependencies
     gitRepoClean.value
+    stageAndCacheKey.value
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: Why don't you just use the task in `VersionInjectorPlugin` directly here? I don't see why you have a new task for this . . .

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 3]
```diff
@@ -40,6 +44,18 @@ object VersionInjectorPlugin extends AutoPlugin {
       "gitDescribe",
       "The description of the current git commit"
     )
+    val gitMRC = TaskKey[String](
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: This name should match the other hash-generating task (`gitSha1`). Maybe `gitLocalSha1`?

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 4]
```diff
@@ -40,6 +44,18 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+      "gitMRC",
+      "Most recent commit in src"
+    )
+    val cacheKey = TaskKey[String](
+      "cacheKey",
+      "cacheKey of project"
+    )
+    val mTask = TaskKey[Seq[String]](
+      "fucksbt",
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I agree with the sentiment, but this should probably be removed. :P

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 5]
```diff
@@ -72,6 +91,14 @@ object VersionInjectorPlugin extends AutoPlugin {
     (gitCommand("describe").!!).trim
   }
 
+  def gitMostRecentCommit(path: File) = {
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Return type?

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 6]
```diff
@@ -72,6 +91,14 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+    gitCommand("log", "-1", "--format=%H", path).!!.trim
+  }
+
+  val gitMRCTask = gitMRC := {
+    gitMostRecentCommit(sourceDirectory.value)
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: This should use the project base (`baseDirectory.value`). It's possible that some other files (like `build.sbt`) will affect the output jar - and the number of extra misses due to this will be fairly small.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 7]
```diff
@@ -117,6 +144,45 @@ object VersionInjectorPlugin extends AutoPlugin {
         gitConfFile
     }
 
+  val injectCacheKeyTask =
+    injectCacheKey <<= (resourceManaged in Compile, organization, name, version, streams, cacheKey) map {
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Is there a reason you can't use `:=` and call `.value` on the dependencies?

I ask because I honestly can't figure out how `<<=` differs from `:=`, and I'd prefer to use `:=` if possible.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 8]
```diff
@@ -117,6 +144,45 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+      // scalastyle: ignore
+      (resourceManaged, org, name, version, s, cacheKeyResult) =>
+        val cacheKeyConfFile = resourceManaged / org / cleanArtifactName(name) / "cacheKey.conf"
+        val cacheKeyContents = "cacheKey: " + "\"" + cacheKeyResult + "\""
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Alternately:
```scala
// Triple-quotes with interpolation:
s"""cacheKey: "$cacheKeyResult""""
```

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala (hunk 7 above)
Comment (ryanai3) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I was just following the style of injectGitTask, and injectArtifactTask. I'm sure I could use :=, but I'm not sure how it would work

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 9]
```diff
@@ -117,6 +144,45 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+        IO.write(cacheKeyConfFile, cacheKeyContents)
+        cacheKeyConfFile
+    }
+
+  lazy val gitMRCs = Def.taskDyn {
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: This should be commented. In particular, you should describe what it's doing (finding  the most recent commits for all local dependencies), and why it needs a dynamic task.

Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: A type on this value would also be good.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 10]
```diff
@@ -14,7 +14,7 @@ object VersionInjectorPlugin extends AutoPlugin {
   object autoImport {
     val injectVersion = TaskKey[Seq[File]](
       "injectVersion",
-      "Generates both the artifact.conf and git.conf version files"
+      "Generates the artifact.conf, git.conf, and cacheKey.conf version files"
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Revert - this doesn't generate the cache key anymore.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 11]
```diff
@@ -117,6 +144,45 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+      case _ => Seq.empty
+    }
+    val stageDir = baseDirectory.value + "/target/universal/stage"
+    val allFiles: Seq[JFile] = Seq("bin", "conf", "lib", "public")
+      .map { dir => stageDir + "/" + dir }
+      .flatMap(getFileList)
+
+    val (filesToGetGitSha, filesToHash) = allFiles partition { f: JFile =>
+      val fileName = f.getName
+      fileName.startsWith("org.allenai.ari-api") ||
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: You should either have a TODO to make this a setting, or go ahead and add a setting for ignored local jars.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 12]
```diff
@@ -117,6 +144,45 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+        fileName.startsWith("org.allenai.solvers-") ||
+        fileName.startsWith("org.allenai.ari-solvers-")
+    }
+    val hashes = filesToHash.map(Hash.apply).map(Hash.toHex).mkString
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: This still relies on filesystem and SBT subproject ordering, both of which could fluctuate. You should put all hash values into a set, sort them, then take the final hash value.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala (hunk 7 above)
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I *think* `<<=` just says "I want the values of all of these keys". You should just be able to call `.value` on all of the keys in the body of `:=`.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala (hunk 9 above)
Comment (ryanai3) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: on injectCacheKeyTask, or on gitMRC's?

Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: `gitMRC` should have a type declaration - it's not clear from reading what it is.

**Comment by ryanai3** [comment]
@jkinkead made changes, passes manual tests

File: build.sbt [hunk 13]
```diff
@@ -36,7 +36,7 @@ addSbtPlugin(
 addSbtPlugin("com.github.jkinkead" % "sbt-scalariform" % "0.1.6")
 
 // Dependency graph visualiztion in SBT console
-addSbtPlugin("com.gilt" % "sbt-dependency-graph-sugar" % "0.7.4")
+addSbtPlugin("com.gilt" % "sbt-dependency-graph-sugar" % "0.7.5-1")
```
Comment (jkinkead) [inline build.sbt]: If this isn't required by your plugin, you may want to remove (or put in a separate PR).

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 14]
```diff
@@ -122,6 +124,8 @@ object DeployPlugin extends AutoPlugin {
     IO.delete((UniversalPlugin.autoImport.stagingDirectory in Universal).value)
   }
 
+  val stageAndCacheKeyTask = stageAndCacheKey := VersionInjectorPlugin.injectCacheKeyTask
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: Remove? This is no longer needed.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 15]
```diff
@@ -40,6 +44,14 @@ object VersionInjectorPlugin extends AutoPlugin {
       "gitDescribe",
       "The description of the current git commit"
     )
+    val gitLocalSha1 = TaskKey[String](
+      "gitLocalSha1",
+      "Most recent commit in src"
+    )
+    val cacheKey = TaskKey[String](
+      "cacheKey",
+      "cacheKey of project"
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: You should expand this a bit (and capitalize the first word for consistency with the other keys).

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 16]
```diff
@@ -72,6 +87,14 @@ object VersionInjectorPlugin extends AutoPlugin {
     (gitCommand("describe").!!).trim
   }
 
+  def gitMostRecentCommit(path: File): String = {
+    gitCommand("log", "-1", "--format=%H", path).!!.trim
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I'd move this into the task def below - it's not used elsewhere, so a separate helper isn't really needed. Maybe add a comment to it, though?

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 17]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
         gitConfFile
     }
 
+  val injectCacheKeyTask = injectCacheKey := {
+    val values = ((resourceManaged in Compile).value.absolutePath, organization.value, name.value, streams.value, cacheKey.value)
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I'd split these out into separate lines.
```scala
val resourceManagedPath = (resourceManaged in Compile).value.absolutePath
// etc.
```

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 18]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+    val (resourceManagedPath, org, nameValue, s, cacheKeyResult) = values
+    val cacheKeyConfFile = new java.io.File(s"$resourceManagedPath/$org/${cleanArtifactName(nameValue)}/cacheKey.conf")
+    s.log.info(s"Generating cacheKey.conf managed resource... (cacheKey: $cacheKeyResult)")
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: More common is to do:
```scala
val logger = streams.value.log
logger.info("stuff")
```

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 19]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
         gitConfFile
     }
 
+  val injectCacheKeyTask = injectCacheKey := {
+    val values = ((resourceManaged in Compile).value.absolutePath, organization.value, name.value, streams.value, cacheKey.value)
+    val (resourceManagedPath, org, nameValue, s, cacheKeyResult) = values
+    val cacheKeyConfFile = new java.io.File(s"$resourceManagedPath/$org/${cleanArtifactName(nameValue)}/cacheKey.conf")
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I'd make `cleanName` a local.
```scala
val cleanName = cleanArtifactName(name.value)
```

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 20]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+    s.log.info(s"Generating cacheKey.conf managed resource... (cacheKey: $cacheKeyResult)")
+
+    val cacheKeyContents = s"""cacheKey: "$cacheKeyResult""""
+    IO.write(cacheKeyConfFile, cacheKeyContents)
+    cacheKeyConfFile
+  }
+
+  //Gives us the git most recent commits for all the local projects that this project depends on
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Space after `//`.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 21]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+  lazy val gitMRCs = Def.taskDyn { // We have to use a dynamic task because generating tasks based on project dependencies
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I'd put this comment up above, with the other comment on the task.

Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Also, you should rename to something clearer, and add a type.
```scala
lazy val dependentGitCommits: Seq[Task[String]] // or whatever this type is
```

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 22]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+    // necessarily alters the task graph
+    val MRCs = buildDependencies.value.classpathRefs(thisProjectRef.value).sorted // get the local dependencies
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: It would be better to sort the SHA1 values, since you know that they have a stable sort order. I'm not sure what the type here is, but it looks like an sbt internal type that may change ordering on you.

Also, you probably just want to sort at the very end - the ordering in this function isn't really important . . .

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 23]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+    val filter = ScopeFilter(inProjects(MRCs: _*)) // this is weird, we create a scopefilter on the dependencies
+    gitLocalSha1.all(filter) // odd piece of syntax - returns a list of tasks (we're inside a taskDyn block) that is
+    //applying gitLocalSHa1 to all scopes in the scopefilter 'filter'
+  }
+
+  val cacheKeyTask = cacheKey := {
+    import java.io.{ File => JFile }
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: I think you can just use sbt's builtin - it has implicit conversions . . .

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 24]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+
+    def getFileList(dir: String): Seq[JFile] = new JFile(dir) match {
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Did you try using the path matchers I linked you to? You should be able to one-line most of this helper . . .

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 25]
```diff
@@ -117,6 +140,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+      //TODO: make this a setting
+      val fileName = f.getName
+      fileName.startsWith("org.allenai.ari-api") ||
+        fileName.startsWith("org.allenai.solvers-") ||
+        fileName.startsWith("org.allenai.ari-solvers-")
+    }
+    val hashes = filesToHash.sorted
+      .map(Hash.apply)
+      .map(Hash.toHex)
+    Hash.toHex(Hash((hashes ++ gitMRCs.value :+ gitLocalSha1.value).sorted.mkString))
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: This is the only place you need to have a `sorted` call. You should add a comment, too, so that the sorting doesn't get deleted by an over-zealous refactorer later on . . .

**Comment by jkinkead** [comment]
More comments!

**Comment by ryanai3** [comment]
More fixes!

File: build.sbt (hunk 13 above)
Comment (jkinkead) [inline build.sbt]: Ping!

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala (hunk 20 above)
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Ping!

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 26]
```diff
@@ -117,6 +136,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+
+    val cacheKeyContents = s"""cacheKey: "$cacheKeyResult""""
+    IO.write(cacheKeyConfFile, cacheKeyContents)
+    cacheKeyConfFile
+  }
+
+  //Gives us the git most recent commits for all the local projects that this project depends on
+  // We have to use a dynamic task because generating tasks based on project dependencies
+  lazy val dependentGitCommits: Def.Initialize[Task[Seq[String]] = Def.taskDyn {
+    // necessarily alters the task graph
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Move up to the first half of this comment.

Also, you should terminate your sentences with a period.

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 27]
```diff
@@ -117,6 +136,51 @@ object VersionInjectorPlugin extends AutoPlugin {
...
+    val MRCs = buildDependencies.value.classpathRefs(thisProjectRef.value)// get the local dependencies
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Prefer comments above the line they're documenting to at the end-of-line.

**Comment by jkinkead** [comment]
Awesome, this is starting to look really great! Let me know when you've fixed up your comments - I think this will be good to merge after that. :)

You should consider working on the TODO in your next PR (making the filtered jars a setting).

**Comment by ryanai3** [comment]
Fixed comments, Merging!

File: build.sbt [hunk 28]
```diff
@@ -36,7 +36,7 @@ addSbtPlugin(
 addSbtPlugin("com.github.jkinkead" % "sbt-scalariform" % "0.1.6")
 
 // Dependency graph visualiztion in SBT console
-addSbtPlugin("com.gilt" % "sbt-dependency-graph-sugar" % "0.7.4")
+addSbtPlugin("com.gilt" % "sbt-dependency-graph-sugar" % "0.7.4444")
```
Comment (jkinkead) [inline build.sbt]: You didn't fix this - is this a valid version?
